flakes/k8s-controller-tools/version.patch

```diff
@@ -0,0 +1,24 @@
+diff --git a/pkg/version/version.go b/pkg/version/version.go
+index 09c8efcf..b9ec798a 100644
+--- a/pkg/version/version.go
++++ b/pkg/version/version.go
+@@ -20,14 +20,12 @@ import (
+ 	"runtime/debug"
+ )
+ 
++var version string
++
+ // Version returns the version of the main module
+ func Version() string {
+-	info, ok := debug.ReadBuildInfo()
+-	if !ok || info == nil || info.Main.Version == "" {
+-		// binary has not been built with module support or doesn't contain a version.
+-		return "(unknown)"
+-	}
+-	return info.Main.Version
```

Comment:
did you include this patch manually? What was the process?

Reply:
This is the same approach as the original nixos package:
https://github.com/NixOS/nixpkgs/tree/nixos-unstable/pkgs/development/tools/kubernetes-controller-tools
You have the `default.nix` file that we converted here to a flake, and then you reference from the flake code the patch like this:

```nix
  patches = [ ./version.patch ];
 ```
 
 Being in the git repo, the nix tool has this file available locally.
 
 BTW I lost quite some time with this. I wrongly assumed this patch was only needed for the older 0.14 version and did not look what it was actually doing.
 
 Then I realised that nix was producing a binary with version `(devel)` which can be explained looking here:
 
 https://github.com/kubernetes-sigs/controller-tools/blob/main/pkg/version/version.go#L37
 
 Took me quite a long time to realise the following:
 - The `controller-tools` use Go's module info to infer the binary version.
 - Go modules have a version as modules, but the main module is always `devel`, it does not have a version.
 - This means that when you install with `go get ...@X.Y.Z` the module will be labeled as version `X.Y.Z` but if you compile it, it will be always `devel`.
 - I suppose this is to protect the module from appearing somewhere with the same version but different code.
 - Both the nix package and our flake now patch the version code to be able to set the version in the usual `ldflags -X` way, but overriding a variable within the binary at compile time.